### PR TITLE
Zend Debugger for PHP

### PR DESCRIPTION
**Pull request by bartlomiej-laczkowski**
### What does this PR do?

Adds support for debugging applications that runs on PHP with Zend Debugger on board.
### What issues does this PR fix or reference?
- PHP Support [#2590](https://github.com/eclipse/che/issues/2590)
### New behavior
- Possibility to add and run new _PHP - ZEND DEBUGGER_ debug configuration in Che
- Possibility to debug local PHP applications in CLI or Web mode

**NOTE:** Detailed description can be found in plug-in README file.

### CODE REVIEW

**Comment by benoitf** [comment]
ci-build

File: plugins/plugin-zend-debugger/.gitignore [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Eclipse  #
```
Comment (benoitf) [inline plugins/plugin-zend-debugger/.gitignore]: We have a root .gitgnore so I'm not sure this file is needed

File: plugins/plugin-zend-debugger/che-plugin-zend-debugger-ide/src/main/java/org/eclipse/che/plugin/zdb/ide/ZendDbgFqnResolver.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Rogue Wave Software, Inc. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.plugin.zdb.ide;
+
+import static com.google.common.base.Preconditions.*;
```
Comment (benoitf) [inline plugins/plugin-zend-debugger/che-plugin-zend-debugger-ide/src/main/java/org/eclipse/che/plugin/zdb/ide/ZendDbgFqnResolver.java]: wildcards imports are not in sync with current style

File: plugins/plugin-zend-debugger/che-plugin-zend-debugger-ide/src/test/java/org/eclipse/che/plugin/zdb/ide/configuration/ZendDbgConfigurationPagePresenterTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Rogue Wave Software, Inc. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.plugin.zdb.ide.configuration;
+
+import static org.eclipse.che.plugin.zdb.ide.configuration.ZendDbgConfigurationType.*;
+import static org.mockito.Matchers.*;
+import static org.mockito.Mockito.*;
```
Comment (benoitf) [inline plugins/plugin-zend-debugger/che-plugin-zend-debugger-ide/src/test/java/org/eclipse/che/plugin/zdb/ide/configuration/ZendDbgConfigurationPagePresenterTest.java]: pls avoid usage of wildcards imports

File: plugins/plugin-zend-debugger/che-plugin-zend-debugger-ide/src/test/java/org/eclipse/che/plugin/zdb/ide/configuration/ZendDbgConfigurationTypeTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Rogue Wave Software, Inc. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.plugin.zdb.ide.configuration;
+
+import static org.junit.Assert.*;
```
Comment (benoitf) [inline plugins/plugin-zend-debugger/che-plugin-zend-debugger-ide/src/test/java/org/eclipse/che/plugin/zdb/ide/configuration/ZendDbgConfigurationTypeTest.java]: wildcard imports

File: plugins/plugin-zend-debugger/che-plugin-zend-debugger-server/src/main/java/org/eclipse/che/plugin/zdb/server/ZendDebuggerFactory.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Rogue Wave Software, Inc. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.plugin.zdb.server;
+
+import static java.util.stream.Collectors.*;
```
Comment (benoitf) [inline plugins/plugin-zend-debugger/che-plugin-zend-debugger-server/src/main/java/org/eclipse/che/plugin/zdb/server/ZendDebuggerFactory.java]: wildcards

File: pom.xml [hunk 6]
```diff
@@ -648,6 +648,16 @@
             </dependency>
             <dependency>
                 <groupId>org.eclipse.che.plugin</groupId>
+                <artifactId>che-plugin-zend-debugger-ide</artifactId>
+                <version>${che.version}</version>
+            </dependency>
+			<dependency>
```
Comment (benoitf) [inline pom.xml]: style indent is incorrect

File: plugins/plugin-zend-debugger/che-plugin-zend-debugger-server/src/test/java/.gitignore [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# Ignore everything in this directory
+*
+# Except this file
+!.gitignore
```
Comment (benoitf) [inline plugins/plugin-zend-debugger/che-plugin-zend-debugger-server/src/test/java/.gitignore]: there is a missing EOL

File: plugins/plugin-zend-debugger/che-plugin-zend-debugger-server/src/main/java/org/eclipse/che/plugin/zdb/server/variables/ZendDbgVariable.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Rogue Wave Software, Inc. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.plugin.zdb.server.variables;
+
+import static org.eclipse.che.plugin.zdb.server.expressions.IDbgDataFacet.Facet.*;
```
Comment (benoitf) [inline plugins/plugin-zend-debugger/che-plugin-zend-debugger-server/src/main/java/org/eclipse/che/plugin/zdb/server/variables/ZendDbgVariable.java]: wildcards

**Comment by tolusha** [comment]
@TylerJewell @bartlomiej-laczkowski 
I will take care

**Comment by tolusha** [comment]
@bartlomiej-laczkowski 
I can't create workspace from `ZEND` stack.

ZEND stack error

**Comment by kaloyan-raev** [comment]
@tolusha It is not clear from this screenshot where the creation of the Zend stack actually fails. Could you give steps to reproduce?

**Comment by tolusha** [comment]
@kaloyan-raev 
I've just tried to create  a workspace from Zend stack.

**Comment by kaloyan-raev** [comment]
@tolusha Please give the _exact_ steps :)

I see you typed 'z' in the Filter box. Does it fail there? Or when you hit the Enter keyboard key? Or if you click on the some of the display tag? Or where? What does "fail" actually means?

We have never seen the Zend stack failing to create. Is the issue you see isolated only to the Zend stack? Did you try the same creation steps for another stack?

**Comment by tolusha** [comment]
@kaloyan-raev 
It seems was some local issue. Now it works fine.

**Comment by tolusha** [comment]
@bartlomiej-laczkowski 
Hi. I was able to run PHP debugger.
So far I have several issues:
- breakpoint isn't activated if I added it before starting debugger
- when I change the value of the variable it isn't updated it the view
- different stack traces in attachment

[che-wksp-peyd 2016-11-02 15-48-46.log.zip](https://github.com/eclipse/che/files/566698/che-wksp-peyd.2016-11-02.15-48-46.log.zip)

Let me know if you need more details.

**Comment by bartlomiej-laczkowski** [comment]
@tolusha 
Hello, referring to the issues that you have found:

> breakpoint isn't activated if I added it before starting debugger

Not sure whether to call it a feature or limitation :) Breakpoints can be activated when debug session is started by Zend Debugger engine (so when some PHP script is executed with some params that triggers debug session). When you start "PHP - Zend Debugger" debugger in Che we just simply start some client that waits for PHP debug session to be started by the engine. Once it is started and we have that info on the client side, we can send some breakpoints to the engine and notify Che that breakpoints were activated. That is why breakpoints are not visible as active just right after starting debugger client in Che, it's just a different story than it is i.e. for Java debugger. I can activate breakpoints right after starting the client but it won't mean that they are really active.

> when I change the value of the variable it isn't updated it the view

I guess that you probably use Zend stack with PHP 7. In that case "change value" feature won't work as there is a bug in Zend Debugger extension for PHP 7 (you should see some "segmentation faults" in terminal). Hopefully it will be solved in the near future. Anyway changing value should work with PHP 5.6.x + Zend Debugger.

> different stack traces in attachment

Looks like some stuff that should be fixed by me.

Thanks for the first review!

**Comment by slemeur** [comment]
Hi @bartlomiej-laczkowski 

Could you please rename, in the Debug Configurations, the debugger to 'PHP [ZEND]' instead of 'PHP - ZEND DEBUGGER' (which is really long). 

Thanks in advance

**Comment by bartlomiej-laczkowski** [comment]
Hi @slemeur 

PHP is having now 3 different debuggers: 'Zend Debugger', 'Xdebug' and 'phpdbg' so the initial idea was to make it 'PHP - debugger name' to differentiate possible future implementations for all of them. I also wanted to use full extension/debugger names as PHP devs are familiar with those names. Anyway I am opened for any suggestions how to unify the naming for different PHP debugger configurations as there can be even three of them. BTW why square brackets? Is it some kind of Codenvy required style?

**Comment by bartlomiej-laczkowski** [comment]
@tolusha 
Hello, I have added a few improvements that should in general fix all of the issues that I could notice in the log file provided by you. I think that the current state is ready for the next review. Thanks in advance!

**Comment by slemeur** [comment]
@bartlomiej-laczkowski Thanks!

In general we try to keep names of components as short as possible. 
I understand that PHP devs are familiar with those names, but in the "Debug Configurations" wizard, having the "PHP [ZEND]" should be self explicit.
The square brackets are a proposal. "PHP - ZEND" would be good as well?

**Comment by kaloyan-raev** [comment]
The name of the debugger is "Zend Debugger". "PHP - ZEND" does not really mean anything.

Perhaps, as we have only one PHP debugger integrated for now, we can just name it "PHP". The configuration page already says that it is about configuration of Zend Debugger. So, hopefully, PHP developers who used to work with Xdebug won't be confused.

When we introduce another PHP debugger in Che, we can discuss again how we can name them.

**Comment by tolusha** [comment]
@bartlomiej-laczkowski 
Thx. I am checking.

**Comment by slemeur** [comment]
Yes, I agree, would be better solution for the time being.
+1 to @kaloyan-raev proposal.

Thanks

**Comment by bartlomiej-laczkowski** [comment]
@slemeur @kaloyan-raev 
Alrighty then, let's make it just 'PHP' for now.

**Comment by tolusha** [comment]
I've faced with some NPEs on logs. But it seems everything works well.

[che-wksp-1tnb 2016-11-09 15-46-23.log.zip](https://github.com/eclipse/che/files/580848/che-wksp-1tnb.2016-11-09.15-46-23.log.zip)

**Comment by tolusha** [comment]
@bartlomiej-laczkowski 
let me know when I can try a new version :)

**Comment by bartlomiej-laczkowski** [comment]
@tolusha 
Hello, you can try it now. I have added two more fixes that should do the job (NPEs while deleting breakpoint and setting variable value).

**Comment by tolusha** [comment]
@bartlomiej-laczkowski 
Nice work.

It is possible to have some tests for server-side?

**Comment by bartlomiej-laczkowski** [comment]
@tolusha 
Thanks! I will apply some basic tests for the server side soon,

**Comment by bartlomiej-laczkowski** [comment]
@tolusha  
Hi Anatoliy, I have added server-side tests. To be able to run them you need to have PHP + Zend Debugger installed on your test/OS environment. Note that the 'skipTests' flag is set to 'true' in server's pom.xml so it also needs to be changed.

All in all I think that everything is ready for the final review.

**Comment by tolusha** [comment]
@bartlomiej-laczkowski 
The only issue I found when I was trying to change value. Could you have a look pls.
```
2016-11-29 10:39:39,534[nio-4401-exec-4]  [ERROR] [o.e.c.p.z.server.ZendDebugger 331]   - Could not get debugger engine response for AssignValueRequest [ID=33]
org.eclipse.che.plugin.zdb.server.exceptions.ZendDbgTimeoutException: Response timeout (10 SECONDS) occurred.
	at org.eclipse.che.plugin.zdb.server.connection.ZendDbgConnection$EngineSyncResponse.get(ZendDbgConnection.java:238) ~[che-plugin-zend-debugger-server-5.0.0-M8-SNAPSHOT.jar:5.0.0-M8-SNAPSHOT]
	at org.eclipse.che.plugin.zdb.server.connection.ZendDbgConnection.sendRequest(ZendDbgConnection.java:329) ~[che-plugin-zend-debugger-server-5.0.0-M8-SNAPSHOT.jar:5.0.0-M8-SNAPSHOT]
	at org.eclipse.che.plugin.zdb.server.expressions.ZendDbgExpressionEvaluator.requestAssignment(ZendDbgExpressionEvaluator.java:311) [che-plugin-zend-debugger-server-5.0.0-M8-SNAPSHOT.jar:5.0.0-M8-SNAPSHOT]
	at org.eclipse.che.plugin.zdb.server.expressions.ZendDbgExpressionEvaluator.assign(ZendDbgExpressionEvaluator.java:279) [che-plugin-zend-debugger-server-5.0.0-M8-SNAPSHOT.jar:5.0.0-M8-SNAPSHOT]
	at org.eclipse.che.plugin.zdb.server.expressions.ZendDbgExpression.setValue(ZendDbgExpression.java:67) [che-plugin-zend-debugger-server-5.0.0-M8-SNAPSHOT.jar:5.0.0-M8-SNAPSHOT]
	at org.eclipse.che.plugin.zdb.server.variables.ZendDbgVariable.setValue(ZendDbgVariable.java:94) [che-plugin-zend-debugger-server-5.0.0-M8-SNAPSHOT.jar:5.0.0-M8-SNAPSHOT]
	at org.eclipse.che.plugin.zdb.server.ZendDebugger.setValue(ZendDebugger.java:313) [che-plugin-zend-debugger-server-5.0.0-M8-SNAPSHOT.jar:5.0.0-M8-SNAPSHOT]
	at org.eclipse.che.api.debugger.server.DebuggerService.setValue(DebuggerService.java:188) [che-core-api-debug-5.0.0-M8-SNAPSHOT.jar:5.0.0-M8-SNAPSHOT]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_65]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_65]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_65]
```

**Comment by bartlomiej-laczkowski** [comment]
@tolusha 
Which version of PHP + Zend Debugger did you use?
If you were using PHP 7.0.x I then there is a bug in debugger engine related to request for changing variable value that causes PHP to crash and that is why we can not get the response. It works ok when trying to change some of top level variables but it will fail i.e. when trying to change some object's field value. Changing value works fine in PHP 5.x, anyway if you were using PHP 7.0.x then there is nothing I can do about it (I have reported the issue to the engine team and hopefully they will add a fix soon).

**Comment by tolusha** [comment]
@bartlomiej-laczkowski 
I created project using ZEND stack.
```
PHP 7.0.11 (cli) (built: Sep 28 2016 11:53:58) ( NTS )                                                                                                                                                                          
Copyright (c) 1997-2016 The PHP Group      
```

**Comment by bartlomiej-laczkowski** [comment]
@tolusha 
So unfortunately you faced this issue. As I mentioned before, we can not do anything about it and we need to wait for the fix in the next version of Zend Debugger extension (we will update Zend stack as soon as this fix will be added).

**Comment by tolusha** [comment]
@bartlomiej-laczkowski 
Ok No prb then.

I've created a new PR with small changes and code formatting. I didn't want to disturb you ^)
https://github.com/eclipse/che/pull/3202

**Comment by tolusha** [comment]
I will close this one if you don't mind.

**Comment by bartlomiej-laczkowski** [comment]
OK, no problem, thanks!

**Pull request closed without merging**